pkgs/applications/virtualization/xen/generic.nix

```diff
@@ -1,265 +1,376 @@
+
+    makeFlags = [
+      "XSM_ENABLE=y"
+      "FLASK_ENABLE=y"
+
+      "EFI_VENDOR=nixos"
+      "LD=${efiBinutils}/bin/ld"
+
+      "PREFIX=$(out)"
+      "CONFIG_DIR=/etc"
+      "XEN_EXTFILES_URL=\\$(XEN_ROOT)/xen_ext_files"
+      "XEN_SCRIPT_DIR=$(CONFIG_DIR)/xen/scripts"
+      "BASH_COMPLETION_DIR=$(PREFIX)/share/bash-completion/completions"
+    ] ++ (config.makeFlags or [ ]);
+
+    preBuild = ''
+      ${config.preBuild or ""}
+    '';
+
+    buildFlags = [
+      "xen"
+      "tools"
     ];
-  } // (config.meta or {});
-} // removeAttrs config [ "xenfiles" "buildInputs" "patches" "postPatch" "meta" ])
+
+    postBuild = ''
+      make -C docs man-pages
+
+      ${withTools "buildPhase" (name: pkg: pkg.buildPhase)}
+    '';
+
+    installPhase =
+      # Create $out directories and copy build output.
+      ''
+        mkdir -p $out $out/share $out/share/man
+        cp -prvd dist/install/nix/store/*/* $out/
+        cp -prvd dist/install/boot $out/boot
+        cp -prvd dist/install/etc $out
+        cp -dR docs/man1 docs/man5 $out/share/man/
+      ''
+
+      # Wrap xencov_split, xenmon and xentrace_format.
+      + ''
+        wrapPythonPrograms
+      ''
+
+      # We also need to wrap pygrub, which lies in $out/lib.
+      + ''
+        wrapPythonProgramsIn "$out/lib" "$out $pythonPath"
+      ''
+
+      # Fix shebangs in Xen's various scripts.
+      + ''
+        shopt -s extglob
+        for i in $out/etc/xen/scripts/!(*.sh); do
+          sed -i "2s@^@export PATH=$out/bin:${scriptEnvPath}\n@" $i
+        done
+      ''
+
+      + "${withTools "installPhase" (name: pkg: pkg.installPhase)}";
+
+    postFixup =
+      # Fix binaries in $out/lib/xen/bin.
+      ''
+        addAutoPatchelfSearchPath $out/lib
+        autoPatchelf $out/lib/xen/bin/
+      '';
+
+    enableParallelBuilding = true;
+
+    #TODO: Stop referencing args here.
+    meta = {
+      branch = branch;
+      description =
+        "Xen Hypervisor"
+        + lib.strings.optionalString (config.prefetchedSources != { }) " and related components"
+        + lib.strings.optionalString (args ? meta && args.meta ? description) " (${args.meta.description})";
+      longDescription =
+        (args.meta.longDescription or "")
+        + lib.strings.optionalString (config.prefetchedSources != { }) "\nIncludes:\n"
+        + withPrefetchedSources (
+          name: pkg:
+          "* ${lib.strings.removePrefix "etherboot/" (lib.strings.removePrefix "firmware/" (lib.strings.removeSuffix "-dir-remote" (lib.strings.removeSuffix ".git" name)))}: ${
+            pkg.meta.description or "(No description)"
+          }."
+        );
+      homepage = "https://xenproject.org/";
+      downloadPage = "https://downloads.xenproject.org/release/xen/${version}/";
+      changelog = "https://wiki.xenproject.org/wiki/Xen_Project_${branch}_Release_Notes";
+      license = lib.licenses.gpl2Only;
+      maintainers = [ lib.maintainers.sigmasquadron ];
+      mainProgram = "xl";
```

Comment:
@SigmaSquadron elaborate why is this main program?

Reply:
Because `xl` is the main way to interface with your domains in a dom0. It's the closest thing Xen has to a primary program. Every other binary is meant to be called in a systemd service or through `xl`.

Reply:
fair